### PR TITLE
POC: 2D EAs via composition

### PR DESCRIPTION
**Pull request by jbrockmendel**
Plenty of kludges and linting errors in here, just want to push it to add composition to the discussion.

Instead of patching existing EAs, this introduces ReshapeableArray which just wraps other EAs, and implements reshape methods.  EAs that _do_ natively support 2D can set a `_allows_2d = True` and avoid being wrapped.

In the process of getting this passing, found a handful of new issues/bugs.  Will try to push fixes for those independently.

### CODE REVIEW

File: pandas/core/arrays/categorical.py [hunk 1]
```diff
@@ -457,11 +460,14 @@ def _formatter(self, boxed=False):
...
         return None
 
-    def copy(self):
+    def copy(self, deep: bool = False):
         """
         Copy constructor.
         """
-        return self._constructor(values=self._codes.copy(),
+        values = self._codes
+        if deep:
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: this is the separate issue you reported yes?

File: pandas/core/algorithms.py [hunk 2]
```diff
@@ -105,6 +105,9 @@ def _ensure_data(values, dtype=None):
         else:
             # Datetime
             from pandas import DatetimeIndex
+            from pandas.core.arrays import unwrap_reshapeable
```
Comment (TomAugspurger) [inline pandas/core/algorithms.py]: Is this for something like `factorize(EA)`? Shouldn't the EA (or your wrapper) do this in `_values_for_factorize`?

File: pandas/core/algorithms.py [hunk 3]
```diff
@@ -1575,7 +1578,11 @@ def take_nd(arr, indexer, axis=0, out=None, fill_value=np.nan, mask_info=None,
     # TODO(EA): Remove these if / elifs as datetimeTZ, interval, become EAs
     # dispatch to internal type takes
     if is_extension_array_dtype(arr):
-        return arr.take(indexer, fill_value=fill_value, allow_fill=allow_fill)
+        try:
+            return arr.take(indexer, fill_value=fill_value, allow_fill=allow_fill, axis=axis)
+        except TypeError:
```
Comment (TomAugspurger) [inline pandas/core/algorithms.py]: I worry a bit about `axis` not being the only reason for a TypeError. Can you check `_allows_2d` instead?

**Review by TomAugspurger — COMMENTED** [review]
So high-level, this makes `Block.values` a ReshapableArray, which is an ExtensionArray implementing the 2-D interface. Then a DataFrame is made up of a collection of Blocks whose values are all reshapable, either by being an ndarray, or an ExtensionArray with `_allows_2d = True`?

Is this your preferred approach for fixing `Block.shape == Block.values.shape` going forward?

File: pandas/core/arrays/categorical.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/categorical.py]: Yes, this is #26995.  This is not orthogonal though, needs to be fixed for this POC to work.

File: pandas/core/algorithms.py [hunk 4]
```diff
@@ -1525,7 +1528,7 @@ def take(arr, indices, axis=0, allow_fill=False, fill_value=None):
 
     if allow_fill:
         # Pandas style, -1 means NA
-        validate_indices(indices, len(arr))
+        validate_indices(indices, arr.shape[axis])
```
Comment (jbrockmendel) [inline pandas/core/algorithms.py]: See #26976.

File: pandas/core/algorithms.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/algorithms.py]: Good idea.

File: pandas/core/algorithms.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/algorithms.py]: values_for_factorize might do it, ill check.  But at this point we dont _necessarily_ have a EA, so need a conditional un-wrapper.

Making DTA validate that inputs are 1D can be done separately from the rest of this, which should resolve this particular part of the diff

**Comment by jbrockmendel** [comment]
> So high-level, this makes Block.values a ReshapableArray, which is an ExtensionArray implementing the 2-D interface. Then a DataFrame is made up of a collection of Blocks whose values are all reshapable, either by being an ndarray, or an ExtensionArray with _allows_2d = True?

Yes.

> Is this your preferred approach for fixing Block.shape == Block.values.shape going forward?

No, this is my second-best.  First-best would be to require EAs to handle the (1, N) case themselves, so we wouldn't need this extra layer.  But I definitely prefer this to the metaclass approach, which I wasn't able to get working at all (MRO issues)

**Pull request closed without merging**